src/responder/gasQueue.ts

```diff
@@ -286,14 +287,55 @@ export class GasQueue {
         return new GasQueue(clonedArray, this.emptyNonce, this.replacementRate, this.maxQueueDepth);
     }
 
-    //TODO:198: documentation and tests
-    public unlock(requests: GasQueueItemRequest[]): GasQueue {
-        // we need to keep track of the max gas price used at a given nonce
-        // since we need to go at least higher than this
+    /**
+     * Returns all queue items that are in this queue but not in the supplied queue
+     * @param otherQueue
+     */
+    public difference(otherQueue: GasQueue): GasQueueItem[] {
+        return this.queueItems.filter(tx => !otherQueue.queueItems.includes(tx));
+    }
+
+    /**
+     * Checks to see if this queue contains an item with the supplied identifier
+     * @param queueItem
+     */
+    public contains(identifier: PisaTransactionIdentifier): boolean {
+        return this.queueItems.findIndex(i => i.request.identifier.equals(identifier)) !== -1;
+    }
+
+    
+    /**
+     * Re-add some items that have lower nonces than any in the current queue
+     * @param lowerNonceItems 
+     */
+    public unlock(lowerNonceItems: GasQueueItem[]): GasQueue {
+        // a correct queue is ordered by both nonce and ideal gas price
+        // We'll need to adjust the nonces of queue items to ensure this
+        const allItemsOrderedByNonce = lowerNonceItems.concat(this.queueItems).sort((a, b) => a.nonce - b.nonce);
+
+        const allItemsByIdealGas = [...allItemsOrderedByNonce].sort((a, b) => {
+            if (a.idealGasPrice.gt(b.idealGasPrice)) return -1;
+            else if (a.idealGasPrice.eq(b.idealGasPrice)) return 0;
+            else return 1;
+        });
+
+        for (let index = 0; index < allItemsByIdealGas.length; index++) {
+            const itemOrderedByGas = allItemsByIdealGas[index];
+            const itemOrderedByNonce = allItemsOrderedByNonce[index];
 
-        // we also need to unlock the specified number of nonces
-        // classify those nonces as empty, and re-arrange
+            // if there's a difference in the items then we have a difference in
+            // order. We replace the item with the gas ordered one so that we have
+            // all items ordered by gas price, and by nonce.
+            if (itemOrderedByGas !== itemOrderedByNonce) {
+                this.replace(allItemsOrderedByNonce, index, itemOrderedByGas.request);
+            }
+        }
+
+        // all items ordered by nonce are now also ordered by gas price
+        return new GasQueue(allItemsOrderedByNonce, this.emptyNonce, this.replacementRate, this.maxQueueDepth);
+    }
 
-        throw new ApplicationError("Not implemented");
+    public getItem(request: GasQueueItemRequest) {
+        const queueItem = this.queueItems.find(q => q.request === request);
     }
```

Comment:
I guess you meant to return `queueItem`, but this method is not used anyway

Reply:
Removed